htlcswitch/link.go

```diff
@@ -2503,12 +2610,56 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
 	}, false)
 }
 
-// fail helper function which is used to encapsulate the action necessary for
-// proper disconnect.
-func (l *channelLink) fail(format string, a ...interface{}) {
+// fail is a function which is used to encapsulate the action necessary for
+// properly failing the link. It takes a linkFailureError, which will be used
+// to determine if we should force close the channel, and if we should send an
+// error message to the remote peer. If we decide to force close, we will in
+// addition to disconnecting the peer unilateraly close the channel.
+func (l *channelLink) fail(linkErr linkFailureError,
+	format string, a ...interface{}) {
 	reason := errors.Errorf(format, a...)
-	log.Error(reason)
-	go l.cfg.Peer.Disconnect(reason)
+	l.errorf("Failing link: %s", reason)
+
+	go func() {
+		// Send an error to the peer, why we failed the channel.
+		if linkErr.shouldSendToPeer() {
+			// If sendData is set, send it to the peer. If not,
+			// we'll use the standard error messages in the
+			// payload. We only include sendData in the cases where
+			// the error data does not contain sensitive
+			// information.
+			data := []byte(linkErr.Error())
+			if linkErr.sendData != nil {
+				data = linkErr.sendData
+			}
+			err := l.cfg.Peer.SendMessage(&lnwire.Error{
+				ChanID: l.ChanID(),
+				Data:   data,
+			}, true)
+			if err != nil {
+				l.errorf("unable to send msg to "+
+					"remote peer: %v", err)
+			}
+		}
+
+		l.cfg.Peer.Disconnect(reason)
+	}()
+
+	if linkErr.forceClose {
+		l.warnf("Force closing link.")
+
+		// If we have decided to force close this channel, we'll wait
+		// for the method to return with a close tx. This ensures that
+		// the channel arbitrator has properly removed the link from
+		// our set of active links, and broadcasted the closing tx,
+		// such that no new updates will be attempted on this channel.
+		closeTx, err := l.cfg.ForceClose()
```

Comment:
In practice, I think this has the possibility to deadlock: we'll call into the `contractcourt`, which will then attempt to make a sync call to the switch to disconnect the link, however we're blocked on this initial call leading to a deadlock. As a result, we may need to run this asynchronously. It should be pretty straight forward to attempt to introduce the deadlock locally: open a channel, then shutdown one node and like change the csv value or something, then attempt to send a payment. The receiving node should then attempt to automatically force close.